File: Week-Wise-Assignments/Assignment 10/Systematic Solutions/Python Programs/2_graph_adjacency_list.py
from collections import deque, defaultdict

class GraphList:
    def __init__(self, vertices=0, directed=False):
        self.vertices = vertices
        self.directed = directed
        self.adj_list = defaultdict(list)
    
    def add_edge(self, src, dest):
        """Add an edge between source and destination vertices"""
        if 0 <= src < self.vertices and 0 <= dest < self.vertices:
            self.adj_list[src].append(dest)
            if not self.directed:
                self.adj_list[dest].append(src)
            print(f"Edge added: {src} -> {dest}")
            if not self.directed:
                print(f"Edge added: {dest} -> {src} (undirected)")
        else:
            print(f"Invalid vertices: {src}, {dest}")
    
    def remove_edge(self, src, dest):
        """Remove an edge between source and destination vertices"""
        if 0 <= src < self.vertices and 0 <= dest < self.vertices:
            if dest in self.adj_list[src]:
                self.adj_list[src].remove(dest)
                print(f"Edge removed: {src} -> {dest}")
            
            if not self.directed and src in self.adj_list[dest]:
                self.adj_list[dest].remove(src)
        else:
            print(f"Invalid vertices: {src}, {dest}")
    
    def count_vertices(self):
        """Count number of vertices in the graph"""
        return self.vertices
    
    def count_edges(self):
        """Count number of edges in the graph"""
        edges = 0
        for vertex in range(self.vertices):
            edges += len(self.adj_list[vertex])
        
        # For undirected graph, each edge is counted twice
        if not self.directed:
            edges //= 2
        
        return edges
    
    def find_adjacent_vertices(self, vertex):
        """Find all adjacent vertices of a given vertex"""
        if 0 <= vertex < self.vertices:
            return list(self.adj_list[vertex])
        else:
            print(f"Invalid vertex: {vertex}")
            return []
    
    def search_vertex(self, vertex):
        """Search if a vertex exists in the graph"""
        return 0 <= vertex < self.vertices
    
    def display_graph(self):
        """Display the adjacency list"""
        print("\n=== Adjacency List ===")
        for i in range(self.vertices):
            adjacent = self.adj_list[i]
            print(f"Vertex {i}: {' -> '.join(map(str, adjacent)) if adjacent else 'NULL'}")
    
    def get_degree(self, vertex):
        """Get degree of a vertex"""
        if 0 <= vertex < self.vertices:
            if self.directed:
                out_degree = len(self.adj_list[vertex])
                in_degree = 0
                for v in range(self.vertices):
                    if vertex in self.adj_list[v]:
                        in_degree += 1
                return in_degree, out_degree
            else:
                degree = len(self.adj_list[vertex])
                return degree
        return None
    
    def graph_analysis(self):
        """Comprehensive graph analysis"""
        print("\n=== Graph Analysis ===")
        print(f"Number of vertices: {self.count_vertices()}")
        print(f"Number of edges: {self.count_edges()}")
        print(f"Graph type: {'Directed' if self.directed else 'Undirected'}")
        
        print(f"\nVertex degrees:")
        for i in range(self.vertices):
            if self.directed:
                in_deg, out_deg = self.get_degree(i)
                print(f"Vertex {i}: In-degree = {in_deg}, Out-degree = {out_deg}")
            else:
                degree = self.get_degree(i)
                print(f"Vertex {i}: Degree = {degree}")
        
        # Memory usage estimation
        total_edges = sum(len(self.adj_list[i]) for i in range(self.vertices))
        memory_usage = self.vertices * 8 + total_edges * 4  # Rough estimation
        print(f"\nMemory usage: ~{memory_usage} bytes")
        print(f"Space complexity: O(V + E) where V = {self.vertices}, E = {self.count_edges()}")
    
    def dfs(self, start_vertex):
        """Depth-First Search traversal"""
        if not (0 <= start_vertex < self.vertices):
            print("Invalid start vertex!")
            return []
        
        visited = [False] * self.vertices
        result = []
        
        def dfs_util(vertex):
            visited[vertex] = True
            result.append(vertex)
            
            for neighbor in self.adj_list[vertex]:
                if not visited[neighbor]:
                    dfs_util(neighbor)
        
        dfs_util(start_vertex)
        return result
    
    def bfs(self, start_vertex):
        """Breadth-First Search traversal"""
        if not (0 <= start_vertex < self.vertices):
            print("Invalid start vertex!")
            return []
        
        visited = [False] * self.vertices
        queue = deque([start_vertex])
        visited[start_vertex] = True
        result = []
        
        while queue:
            vertex = queue.popleft()
            result.append(vertex)
            
            for neighbor in self.adj_list[vertex]:
                if not visited[neighbor]:
                    visited[neighbor] = True
                    queue.append(neighbor)
        
        return result
    
    def has_cycle(self):
        """Check if the graph has a cycle"""
        if self.directed:
            return self._has_cycle_directed()
        else:
            return self._has_cycle_undirected()
    
    def _has_cycle_directed(self):
        """Check cycle in directed graph using DFS"""
        WHITE, GRAY, BLACK = 0, 1, 2
        color = [WHITE] * self.vertices
        
        def dfs_cycle(vertex):
            if color[vertex] == GRAY:
                return True
            if color[vertex] == BLACK:
                return False
            
            color[vertex] = GRAY
            for neighbor in self.adj_list[vertex]:
                if dfs_cycle(neighbor):
                    return True
            
            color[vertex] = BLACK
            return False
        
        for vertex in range(self.vertices):
            if color[vertex] == WHITE:
                if dfs_cycle(vertex):
                    return True
        return False
    
    def _has_cycle_undirected(self):
        """Check cycle in undirected graph using DFS"""
        visited = [False] * self.vertices
        
        def dfs_cycle(vertex, parent):
            visited[vertex] = True
            
            for neighbor in self.adj_list[vertex]:
                if not visited[neighbor]:
                    if dfs_cycle(neighbor, vertex):
                        return True
                elif neighbor != parent:
                    return True
            
            return False
        
        for vertex in range(self.vertices):
            if not visited[vertex]:
                if dfs_cycle(vertex, -1):
                    return True
        return False
    
    def is_connected(self):
        """Check if the graph is connected (for undirected graphs)"""
        if self.directed:
            print("Connectivity check is for undirected graphs only")
            return False
        
        if self.vertices == 0:
            return True
        
        # Start DFS from vertex 0
        visited = [False] * self.vertices
        
        def dfs_connect(vertex):
            visited[vertex] = True
            for neighbor in self.adj_list[vertex]:
                if not visited[neighbor]:
                    dfs_connect(neighbor)
        
        dfs_connect(0)
        
        # Check if all vertices are visited
        return all(visited)
    
    def create_graph_interactive(self):
        """Create graph interactively"""
        print("\n=== Creating Graph ===")
        try:
            vertices = int(input("Enter number of vertices: "))
            if vertices <= 0:
                print("Invalid number of vertices!")
                return
            
            directed = input("Is the graph directed? (y/n): ").lower() == 'y'
            
            self.vertices = vertices
            self.directed = directed
            self.adj_list = defaultdict(list)
            
            edges = int(input("Enter number of edges: "))
            
            print("Enter edges (source destination):")
            for i in range(edges):
                src, dest = map(int, input(f"Edge {i + 1}: ").split())
                self.add_edge(src, dest)
            
            print("Graph created successfully!")
            
        except ValueError:
            print("Invalid input! Please enter integers only.")

def main():
    graph = GraphList()
    
    print("=== Graph Implementation (Adjacency List) ===")
    print("Adjacency List: Linked list representation")
    print("Space Complexity: O(V + E), Efficient for sparse graphs\n")
    
    while True:
        print("\n--- Menu ---")
        print("1. Create Graph")
        print("2. Add Edge")
        print("3. Remove Edge")
        print("4. Count Vertices")
        print("5. Count Edges")
        print("6. Find Adjacent Vertices")
        print("7. Search Vertex")
        print("8. Display Graph")
        print("9. Graph Analysis")
        print("10. DFS Traversal")
        print("11. BFS Traversal")
        print("12. Check Cycle")
        print("13. Check Connectivity")
        print("14. Exit")
        
        try:
            choice = int(input("Enter your choice: "))
            
            if choice == 1:
                graph.create_graph_interactive()
            
            elif choice == 2:
                src, dest = map(int, input("Enter source and destination vertices: ").split())
                graph.add_edge(src, dest)
            
            elif choice == 3:
                src, dest = map(int, input("Enter source and destination vertices: ").split())
                graph.remove_edge(src, dest)
            
            elif choice == 4:
                count = graph.count_vertices()
                print(f"Number of vertices: {count}")
            
            elif choice == 5:
                count = graph.count_edges()
                print(f"Number of edges: {count}")
            
            elif choice == 6:
                vertex = int(input("Enter vertex: "))
                adjacent = graph.find_adjacent_vertices(vertex)
                print(f"Adjacent vertices of {vertex}: {adjacent}")
            
            elif choice == 7:
                vertex = int(input("Enter vertex to search: "))
                exists = graph.search_vertex(vertex)
                print(f"Vertex {vertex} {'exists' if exists else 'does not exist'} in the graph")
            
            elif choice == 8:
                graph.display_graph()
            
            elif choice == 9:
                graph.graph_analysis()
            
            elif choice == 10:
                start = int(input("Enter starting vertex for DFS: "))
                result = graph.dfs(start)
                print(f"DFS traversal starting from vertex {start}: {result}")
            
            elif choice == 11:
                start = int(input("Enter starting vertex for BFS: "))
                result = graph.bfs(start)
                print(f"BFS traversal starting from vertex {start}: {result}")
            
            elif choice == 12:
                has_cycle = graph.has_cycle()
                print(f"Graph has cycle: {'Yes' if has_cycle else 'No'}")
            
            elif choice == 13:
                is_connected = graph.is_connected()
                print(f"Graph is connected: {'Yes' if is_connected else 'No'}")
            
            elif choice == 14:
                print("Exiting program...")
                break
            
            else:
                print("Invalid choice! Please try again.")
        
        except ValueError:
            print("Invalid input! Please enter a valid number.")
        except KeyboardInterrupt:
            print("\nProgram interrupted. Exiting...")
            break

if __name__ == "__main__":
    main()

File: Week-Wise-Assignments/Assignment 10/Systematic Solutions/C Programs/2_graph_adjacency_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int vertex;
    struct Node* next;
} Node;

typedef struct {
    Node** adjList;
    int vertices;
    int directed;
} Graph;

Node* createNode(int vertex) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) {
        printf("Memory allocation failed!\n");
        return NULL;
    }
    newNode->vertex = vertex;
    newNode->next = NULL;
    return newNode;
}

void initGraph(Graph* g, int vertices, int directed) {
    g->vertices = vertices;
    g->directed = directed;
    
    g->adjList = (Node**)malloc(vertices * sizeof(Node*));
    if (g->adjList == NULL) {
        printf("Memory allocation failed!\n");
        return;
    }
    
    // Initialize all adjacency lists as empty
    for (int i = 0; i < vertices; i++) {
        g->adjList[i] = NULL;
    }
    
    printf("Graph initialized with %d vertices (%s)\n", 
           vertices, directed ? "Directed" : "Undirected");
}

void addEdge(Graph* g, int src, int dest) {
    if (src >= 0 && src < g->vertices && dest >= 0 && dest < g->vertices) {
        // Add edge from src to dest
        Node* newNode = createNode(dest);
        if (newNode != NULL) {
            newNode->next = g->adjList[src];
            g->adjList[src] = newNode;
            printf("Edge added: %d -> %d\n", src, dest);
        }
        
        // For undirected graph, add edge from dest to src
        if (!g->directed) {
            Node* newNode2 = createNode(src);
            if (newNode2 != NULL) {
                newNode2->next = g->adjList[dest];
                g->adjList[dest] = newNode2;
                printf("Edge added: %d -> %d (undirected)\n", dest, src);
            }
        }
    } else {
        printf("Invalid vertices: %d, %d\n", src, dest);
    }
}

void removeEdge(Graph* g, int src, int dest) {
    if (src >= 0 && src < g->vertices && dest >= 0 && dest < g->vertices) {
        // Remove edge from src to dest
        Node* current = g->adjList[src];
        Node* prev = NULL;
        
        while (current != NULL && current->vertex != dest) {
            prev = current;
            current = current->next;
        }
        
        if (current != NULL) {
            if (prev == NULL) {
                g->adjList[src] = current->next;
            } else {
                prev->next = current->next;
            }
            free(current);
            printf("Edge removed: %d -> %d\n", src, dest);
        }
        
        // For undirected graph, remove edge from dest to src
        if (!g->directed) {
            current = g->adjList[dest];
            prev = NULL;
            
            while (current != NULL && current->vertex != src) {
                prev = current;
                current = current->next;
            }
            
            if (current != NULL) {
                if (prev == NULL) {
                    g->adjList[dest] = current->next;
                } else {
                    prev->next = current->next;
                }
                free(current);
            }
        }
    } else {
        printf("Invalid vertices: %d, %d\n", src, dest);
    }
}

int countVertices(Graph* g) {
    return g->vertices;
}

int countEdges(Graph* g) {
    int edges = 0;
    
    for (int i = 0; i < g->vertices; i++) {
        Node* current = g->adjList[i];
        while (current != NULL) {
            edges++;
            current = current->next;
        }
    }
    
    // For undirected graph, each edge is counted twice
    if (!g->directed) {
        edges /= 2;
    }
    
    return edges;
}

void findAdjacentVertices(Graph* g, int vertex) {
    if (vertex < 0 || vertex >= g->vertices) {
        printf("Invalid vertex: %d\n", vertex);
        return;
    }
    
    printf("Adjacent vertices of %d: ", vertex);
    Node* current = g->adjList[vertex];
    
    if (current == NULL) {
        printf("None");
    } else {
        while (current != NULL) {
            printf("%d ", current->vertex);
            current = current->next;
        }
    }
    printf("\n");
}

int searchVertex(Graph* g, int vertex) {
    if (vertex >= 0 && vertex < g->vertices) {
        return 1;  // Vertex exists
    }
    return 0;  // Vertex doesn't exist
}

void displayGraph(Graph* g) {
    printf("\n=== Adjacency List ===\n");
    for (int i = 0; i < g->vertices; i++) {
        printf("Vertex %d: ", i);
        Node* current = g->adjList[i];
        
        if (current == NULL) {
            printf("NULL");
        } else {
            while (current != NULL) {
                printf("%d", current->vertex);
                if (current->next != NULL) {
                    printf(" -> ");
                }
                current = current->next;
            }
        }
        printf("\n");
    }
}

void graphAnalysis(Graph* g) {
    printf("\n=== Graph Analysis ===\n");
    printf("Number of vertices: %d\n", countVertices(g));
    printf("Number of edges: %d\n", countEdges(g));
    printf("Graph type: %s\n", g->directed ? "Directed" : "Undirected");
    
    // Calculate degree for each vertex
    printf("\nVertex degrees:\n");
    for (int i = 0; i < g->vertices; i++) {
        int outDegree = 0;
        int inDegree = 0;
        
        // Calculate out-degree
        Node* current = g->adjList[i];
        while (current != NULL) {
            outDegree++;
            current = current->next;
        }
        
        // Calculate in-degree
        for (int j = 0; j < g->vertices; j++) {
            Node* temp = g->adjList[j];
            while (temp != NULL) {
                if (temp->vertex == i) {
                    inDegree++;
                }
                temp = temp->next;
            }
        }
        
        if (g->directed) {
            printf("Vertex %d: In-degree = %d, Out-degree = %d\n", 
                   i, inDegree, outDegree);
        } else {
            printf("Vertex %d: Degree = %d\n", i, outDegree);
        }
    }
    
    // Memory usage estimation
    int nodeCount = 0;
    for (int i = 0; i < g->vertices; i++) {
        Node* current = g->adjList[i];
        while (current != NULL) {
            nodeCount++;
            current = current->next;
        }
    }
    
    int memoryUsage = g->vertices * sizeof(Node*) + nodeCount * sizeof(Node);
    printf("\nMemory usage: ~%d bytes\n", memoryUsage);
    printf("Space complexity: O(V + E) where V = %d, E = %d\n", 
           g->vertices, countEdges(g));
}

void createGraphInteractive(Graph* g) {
    int vertices, directed, edges, src, dest;
    
    printf("\n=== Creating Graph ===\n");
    printf("Enter number of vertices: ");
    scanf("%d", &vertices);
    
    if (vertices <= 0) {
        printf("Invalid number of vertices!\n");
        return;
    }
    
    printf("Is the graph directed? (1 for Yes, 0 for No): ");
    scanf("%d", &directed);
    
    // Free existing graph if any
    if (g->adjList != NULL) {
        for (int i = 0; i < g->vertices; i++) {
            Node* current = g->adjList[i];
            while (current != NULL) {
                Node* temp = current;
                current = current->next;
                free(temp);
            }
        }
        free(g->adjList);
    }
    
    initGraph(g, vertices, directed);
    
    printf("Enter number of edges: ");
    scanf("%d", &edges);
    
    printf("Enter edges (source destination):\n");
    for (int i = 0; i < edges; i++) {
        printf("Edge %d: ", i + 1);
        scanf("%d %d", &src, &dest);
        addEdge(g, src, dest);
    }
    
    printf("Graph created successfully!\n");
}

void dfsUtil(Graph* g, int vertex, int visited[]) {
    visited[vertex] = 1;
    printf("%d ", vertex);
    
    Node* current = g->adjList[vertex];
    while (current != NULL) {
        if (!visited[current->vertex]) {
            dfsUtil(g, current->vertex, visited);
        }
        current = current->next;
    }
}

void dfs(Graph* g, int startVertex) {
    if (startVertex < 0 || startVertex >= g->vertices) {
        printf("Invalid start vertex!\n");
        return;
    }
    
    int* visited = (int*)calloc(g->vertices, sizeof(int));
    if (visited == NULL) {
        printf("Memory allocation failed!\n");
        return;
    }
    
    printf("DFS traversal starting from vertex %d: ", startVertex);
    dfsUtil(g, startVertex, visited);
    printf("\n");
    
    free(visited);
}

void bfs(Graph* g, int startVertex) {
    if (startVertex < 0 || startVertex >= g->vertices) {
        printf("Invalid start vertex!\n");
        return;
    }
    
    int* visited = (int*)calloc(g->vertices, sizeof(int));
    int* queue = (int*)malloc(g->vertices * sizeof(int));
    
    if (visited == NULL || queue == NULL) {
        printf("Memory allocation failed!\n");
        return;
    }
    
    int front = 0, rear = 0;
    
    visited[startVertex] = 1;
    queue[rear++] = startVertex;
    
    printf("BFS traversal starting from vertex %d: ", startVertex);
    
    while (front < rear) {
        int vertex = queue[front++];
        printf("%d ", vertex);
        
        Node* current = g->adjList[vertex];
        while (current != NULL) {
            if (!visited[current->vertex]) {
                visited[current->vertex] = 1;
                queue[rear++] = current->vertex;
            }
            current = current->next;
        }
    }
    printf("\n");
    
    free(visited);
    free(queue);
}

void freeGraph(Graph* g) {
    if (g->adjList != NULL) {
        for (int i = 0; i < g->vertices; i++) {
            Node* current = g->adjList[i];
            while (current != NULL) {
                Node* temp = current;
                current = current->next;
                free(temp);
            }
        }
        free(g->adjList);
        g->adjList = NULL;
    }
}

int main() {
    Graph g = {NULL, 0, 0};
    int choice, vertex, src, dest, result;
    
    printf("=== Graph Implementation (Adjacency List) ===\n");
    printf("Adjacency List: Linked list representation\n");
    printf("Space Complexity: O(V + E), Efficient for sparse graphs\n\n");
    
    while (1) {
        printf("\n--- Menu ---\n");
        printf("1. Create Graph\n");
        printf("2. Add Edge\n");
        printf("3. Remove Edge\n");
        printf("4. Count Vertices\n");
        printf("5. Count Edges\n");
        printf("6. Find Adjacent Vertices\n");
        printf("7. Search Vertex\n");
        printf("8. Display Graph\n");
        printf("9. Graph Analysis\n");
        printf("10. DFS Traversal\n");
        printf("11. BFS Traversal\n");
        printf("12. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);
        
        switch (choice) {
            case 1:
                createGraphInteractive(&g);
                break;
                
            case 2:
                printf("Enter source and destination vertices: ");
                scanf("%d %d", &src, &dest);
                addEdge(&g, src, dest);
                break;
                
            case 3:
                printf("Enter source and destination vertices: ");
                scanf("%d %d", &src, &dest);
                removeEdge(&g, src, dest);
                break;
                
            case 4:
                result = countVertices(&g);
                printf("Number of vertices: %d\n", result);
                break;
                
            case 5:
                result = countEdges(&g);
                printf("Number of edges: %d\n", result);
                break;
                
            case 6:
                printf("Enter vertex: ");
                scanf("%d", &vertex);
                findAdjacentVertices(&g, vertex);
                break;
                
            case 7:
                printf("Enter vertex to search: ");
                scanf("%d", &vertex);
                result = searchVertex(&g, vertex);
                printf("Vertex %d %s in the graph\n", 
                       vertex, result ? "exists" : "does not exist");
                break;
                
            case 8:
                displayGraph(&g);
                break;
                
            case 9:
                graphAnalysis(&g);
                break;
                
            case 10:
                printf("Enter starting vertex for DFS: ");
                scanf("%d", &vertex);
                dfs(&g, vertex);
                break;
                
            case 11:
                printf("Enter starting vertex for BFS: ");
                scanf("%d", &vertex);
                bfs(&g, vertex);
                break;
                
            case 12:
                printf("Freeing memory and exiting...\n");
                freeGraph(&g);
                exit(0);
                
            default:
                printf("Invalid choice! Please try again.\n");
        }
    }
    
    return 0;
}